src/rgw/driver/rados/rgw_reshard.cc

```diff
@@ -1215,36 +1342,83 @@ int RGWReshard::process_entry(const cls_rgw_reshard_entry& entry,
+      return clean_up("bucket does not currently exist");
     } else {
-      ldpp_dout(dpp, 0) << __func__ <<
-          ": Bucket: " << entry.bucket_name <<
-          " already resharded by someone, skipping " << dendl;
+      return clean_up("bucket already resharded");
     }
+  }
 
-    // we've encountered a reshard queue entry for an apparently
-    // non-existent bucket; let's try to recover by cleaning up
-    ldpp_dout(dpp, 0) <<  __func__ <<
-        ": removing reshard queue entry for a resharded or non-existent bucket" <<
-        entry.bucket_name << dendl;
+  // if reshard reduction, perform extra sanity checks in part to
+  // prevent chasing constantly changing entry count
+  if (entry.new_num_shards < entry.old_num_shards) {
+    const bool may_reduce =
+      store->ctx()->_conf.get_val<bool>("rgw_dynamic_resharding_may_reduce");
+    if (! may_reduce) {
+      return clean_up("current configuration does not allow reshard reduction");
+    }
 
-    ret = remove(dpp, entry, y);
-    if (ret < 0) {
-      ldpp_dout(dpp, 0) << __func__ <<
-          ": Error removing non-existent bucket " <<
-          entry.bucket_name << " from resharding queue: " <<
-          cpp_strerror(-ret) << dendl;
-      return ret;
+    // determine how many entries there are in the bucket index
+    std::map<RGWObjCategory, RGWStorageStats> stats;
+    ret = store->getRados()->get_bucket_stats(dpp, bucket_info,
+					      bucket_info.layout.current_index,
+					      -1, nullptr, nullptr, stats, nullptr, nullptr);
+
+    // determine current number of bucket entries across shards
+    uint64_t num_entries = 0;
+    for (const auto& s : stats) {
+      num_entries += s.second.num_objects;
     }
 
-    // we cleaned up, move on to the next entry
-    return 0;
+    const uint32_t current_shard_count =
+      rgw::num_shards(bucket_info.get_current_index().layout.normal);
+
+    bool needs_resharding { false };
+    uint32_t suggested_shard_count { 0 };
+    // calling this rados function determines various rados values
+    // needed to perform the calculation before calling
+    // calculating_preferred_shards() in this class
+    store->getRados()->calculate_preferred_shards(
+      dpp, num_entries, current_shard_count,
+      needs_resharding, &suggested_shard_count);
+
+    // if we no longer need resharding or currently need to expand
+    // number of shards, drop this request
+    if (!needs_resharding || suggested_shard_count > current_shard_count) {
+      return clean_up("reshard reduction no longer appropriate");
+    }
+
+    // see if it's been long enough since this reshard queue entry was
+    // added to actually do the reshard reduction
+    ceph::real_time when_queued = entry.time;
+    ceph::real_time now = real_clock::now();
+
+    // convert hours to seconds
+    const uint32_t reshard_reduction_wait_period_hours =
+      uint32_t(store->ctx()->_conf.get_val<uint64_t>("rgw_dynamic_resharding_reduce_wait"));
```

Comment:
So we would have to distinguish between a queue entry added via dynamic resharding and one via radosgw-admin. I was trying not to touch the at-rest data structure, but the most straight-forward way would be to add a field.

Reply:
i guess an alternative is to print an error message on `reshard add` recommending manual reshard with `bucket reshard` instead? or at least a warning that the reshard will be delayed by rgw_dynamic_resharding_reduce_wait

Reply:
The two obvious choices would be a timestamp to indicate when the resharding can take place. Or alternatively an enum indicating the source of the request -- currently only _admin_ or _dynamic resharding_. I'm leaning towards the latter in case the admin updates the config while an entry sits on the queue.

Reply:
see below for details of the reshard initiator field to handle this